### ISSUE TITLE
[Phase 1] Feature Engineering Pipeline

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/demand-forecasting-engine#2]
## Summary

Implement core feature engineering transformers that work with sklearn Pipeline for demand forecasting. This includes lag features, rolling statistics, and calendar effects that form the foundation of time-series feature engineering.

## System Context

```
demand-forecasting-engine/
├── src/
│   └── features/
│       ├── __init__.py
│       ├── lag_features.py      # <- CREATE
│       ├── rolling_features.py  # <- CREATE
│       ├── calendar_features.py # <- CREATE
│       └── pipeline.py          # <- CREATE
└── tests/
    └── test_features.py         # <- CREATE
```

## Dependencies

- Requires: Phase 0 (Data Foundation) completed
- Blocks: Phase 2 (Single Model Baseline)

### 1. Lag Feature Transformer (`src/features/lag_features.py`)

```python
import pandas as pd
import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from typing import List


class LagFeatureTransformer(BaseEstimator, TransformerMixin):
    """
    Create lag features for time series data.

    Parameters
    ----------
    lags : list of int
        Lag periods to create (e.g., [1, 7, 14, 28] for daily data)
    target_column : str
        Column to create lags from
    group_columns : list of str, optional
        Columns to group by before creating lags (e.g., ['sku_id'])
    fill_value : float, optional
        Value to fill NaN resulting from lag operation
    """

    def __init__(
        self,
        lags: List[int] = [1, 7, 14, 28],
        target_column: str = 'sales',
        group_columns: List[str] = None,
        fill_value: float = None,
    ):
        self.lags = lags
        self.target_column = target_column
        self.group_columns = group_columns
        self.fill_value = fill_value

    def fit(self, X: pd.DataFrame, y=None):
        """No fitting required for lag features."""
        self.feature_names_out_ = [
            f"{self.target_column}_lag_{lag}" for lag in self.lags
        ]
        return self

    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
        """Create lag features."""
        X = X.copy()

        for lag in self.lags:
            col_name = f"{self.target_column}_lag_{lag}"

            if self.group_columns:
                X[col_name] = X.groupby(self.group_columns)[self.target_column].shift(lag)
            else:
                X[col_name] = X[self.target_column].shift(lag)

            if self.fill_value is not None:
                X[col_name] = X[col_name].fillna(self.fill_value)

        return X

    def get_feature_names_out(self, input_features=None) -> List[str]:
        """Return feature names for output."""
        return self.feature_names_out_
```

### 2. Rolling Statistics Transformer (`src/features/rolling_features.py`)

```python
import pandas as pd
import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from typing import List, Callable


class RollingStatsTransformer(BaseEstimator, TransformerMixin):
    """
    Create rolling window statistics for time series.

    Parameters
    ----------
    windows : list of int
        Rolling window sizes (e.g., [7, 14, 28])
    target_column : str
        Column to compute statistics on
    statistics : list of str
        Statistics to compute: 'mean', 'std', 'min', 'max', 'median'
    group_columns : list of str, optional
        Columns to group by before rolling
    min_periods : int
        Minimum observations required for valid result
    """

    STAT_FUNCS = {
        'mean': lambda x: x.mean(),
        'std': lambda x: x.std(),
        'min': lambda x: x.min(),
        'max': lambda x: x.max(),
        'median': lambda x: x.median(),
    }

    def __init__(
        self,
        windows: List[int] = [7, 14, 28],
        target_column: str = 'sales',
        statistics: List[str] = ['mean', 'std'],
        group_columns: List[str] = None,
        min_periods: int = 1,
    ):
        self.windows = windows
        self.target_column = target_column
        self.statistics = statistics
        self.group_columns = group_columns
        self.min_periods = min_periods

    def fit(self, X: pd.DataFrame, y=None):
        """Generate feature names."""
        self.feature_names_out_ = [
            f"{self.target_column}_rolling_{stat}_{window}"
            for window in self.windows
            for stat in self.statistics
        ]
        return self

    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
        """Compute rolling statistics."""
        X = X.copy()

        for window in self.windows:
            for stat in self.statistics:
                col_name = f"{self.target_column}_rolling_{stat}_{window}"

                if self.group_columns:
                    rolling = X.groupby(self.group_columns)[self.target_column].rolling(
                        window=window, min_periods=self.min_periods
                    )
                    X[col_name] = rolling.apply(self.STAT_FUNCS[stat]).reset_index(level=0, drop=True)
                else:
                    rolling = X[self.target_column].rolling(
                        window=window, min_periods=self.min_periods
                    )
                    X[col_name] = self.STAT_FUNCS[stat](rolling)

        return X

    def get_feature_names_out(self, input_features=None) -> List[str]:
        return self.feature_names_out_
```

### 3. Calendar Feature Transformer (`src/features/calendar_features.py`)

```python
import pandas as pd
import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
from typing import List, Optional
import holidays


class CalendarFeatureTransformer(BaseEstimator, TransformerMixin):
    """
    Create calendar-based features from date column.

    Parameters
    ----------
    date_column : str
        Name of the date column
    country : str
        Country code for holidays (e.g., 'US', 'UK')
    include_holidays : bool
        Whether to add holiday features
    """

    def __init__(
        self,
        date_column: str = 'date',
        country: str = 'US',
        include_holidays: bool = True,
    ):
        self.date_column = date_column
        self.country = country
        self.include_holidays = include_holidays

    def fit(self, X: pd.DataFrame, y=None):
        """Initialize holiday calendar."""
        if self.include_holidays:
            years = X[self.date_column].dt.year.unique()
            self.holiday_calendar_ = holidays.country_holidays(
                self.country, years=years.tolist()
            )

        self.feature_names_out_ = [
            'day_of_week', 'day_of_month', 'day_of_year',
            'week_of_year', 'month', 'quarter', 'year',
            'is_weekend', 'is_month_start', 'is_month_end',
        ]
        if self.include_holidays:
            self.feature_names_out_.extend(['is_holiday', 'days_to_holiday', 'days_from_holiday'])

        return self

    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
        """Extract calendar features."""
        X = X.copy()
        date = X[self.date_column]

        # Basic calendar features
        X['day_of_week'] = date.dt.dayofweek
        X['day_of_month'] = date.dt.day
        X['day_of_year'] = date.dt.dayofyear
        X['week_of_year'] = date.dt.isocalendar().week.astype(int)
        X['month'] = date.dt.month
        X['quarter'] = date.dt.quarter
        X['year'] = date.dt.year

        # Boolean features
        X['is_weekend'] = date.dt.dayofweek.isin([5, 6]).astype(int)
        X['is_month_start'] = date.dt.is_month_start.astype(int)
        X['is_month_end'] = date.dt.is_month_end.astype(int)

        # Holiday features
        if self.include_holidays:
            X['is_holiday'] = date.apply(lambda d: d in self.holiday_calendar_).astype(int)
            X['days_to_holiday'] = self._days_to_nearest_holiday(date, direction='forward')
            X['days_from_holiday'] = self._days_to_nearest_holiday(date, direction='backward')

        return X

    def _days_to_nearest_holiday(self, dates: pd.Series, direction: str) -> pd.Series:
        """Calculate days to/from nearest holiday."""
        holiday_dates = sorted(self.holiday_calendar_.keys())

        def find_nearest(d):
            if direction == 'forward':
                future = [h for h in holiday_dates if h >= d.date()]
                return (future[0] - d.date()).days if future else 365
            else:
                past = [h for h in holiday_dates if h <= d.date()]
                return (d.date() - past[-1]).days if past else 365

        return dates.apply(find_nearest)

    def get_feature_names_out(self, input_features=None) -> List[str]:
        return self.feature_names_out_
```

### 4. Feature Pipeline (`src/features/pipeline.py`)

```python
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from .lag_features import LagFeatureTransformer
from .rolling_features import RollingStatsTransformer
from .calendar_features import CalendarFeatureTransformer


def build_feature_pipeline(
    target_column: str = 'sales',
    group_columns: list = ['sku_id'],
    lags: list = [1, 7, 14, 28],
    rolling_windows: list = [7, 14, 28],
    country: str = 'US',
) -> Pipeline:
    """
    Build the complete feature engineering pipeline.

    Returns a sklearn Pipeline that transforms raw sales data
    into ML-ready features.
    """
    pipeline = Pipeline([
        ('lag_features', LagFeatureTransformer(
            lags=lags,
            target_column=target_column,
            group_columns=group_columns,
        )),
        ('rolling_stats', RollingStatsTransformer(
            windows=rolling_windows,
            target_column=target_column,
            group_columns=group_columns,
        )),
        ('calendar', CalendarFeatureTransformer(
            country=country,
        )),
    ])

    return pipeline
```

## Implementation Checklist

- [ ] Create `LagFeatureTransformer` with group-aware lag creation
- [ ] Create `RollingStatsTransformer` with multiple statistics
- [ ] Create `CalendarFeatureTransformer` with holiday support
- [ ] Create `build_feature_pipeline()` function
- [ ] Add `get_feature_names_out()` to all transformers
- [ ] Write unit tests for each transformer
- [ ] Write integration test for full pipeline

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/features/lag_features.py` | 0-80 | Create | Lag feature transformer |
| `src/features/rolling_features.py` | 0-90 | Create | Rolling statistics transformer |
| `src/features/calendar_features.py` | 0-100 | Create | Calendar and holiday features |
| `src/features/pipeline.py` | 0-50 | Create | Pipeline orchestration |
| `tests/test_features.py` | 0-150 | Create | Unit and integration tests |

## Technical Challenges

**Group-aware operations**: When creating lags per SKU, must group data correctly:

```python
# WRONG - lags cross SKU boundaries
df['lag_1'] = df['sales'].shift(1)

# CORRECT - lags respect SKU groups
df['lag_1'] = df.groupby('sku_id')['sales'].shift(1)
```

**NaN handling**: Lag and rolling features create NaN at series start. Strategy:
- Option 1: Drop rows with NaN (loses warmup period)
- Option 2: Fill with 0 or mean (introduces bias)
- Option 3: Use `min_periods=1` for rolling (preferred for rolling stats)

## Definition of Done

- [ ] `LagFeatureTransformer.transform()` creates correct lag features
- [ ] `RollingStatsTransformer.transform()` computes correct statistics
- [ ] `CalendarFeatureTransformer.transform()` adds all calendar features
- [ ] `build_feature_pipeline()` chains all transformers
- [ ] All transformers implement `get_feature_names_out()`
- [ ] Tests pass: `pytest tests/test_features.py -v`

## Risk Assessment

| Risk | Impact | Mitigation |
|------|--------|------------|
| Data leakage via look-ahead | :red_circle: High | Only use `.shift()` with positive values |
| Memory explosion with many lags | :yellow_circle: Medium | Limit to essential lags only |
| Holiday library missing dates | :green_circle: Low | Default to is_holiday=False |

## Time Estimate

**Estimated Hours**: 14-18 hours
**Priority**: Critical (required for model training)
